### PR TITLE
Adding a new delegation strategy with test cases based on mainnet.

### PR DESCRIPTION
**Pull request by arhamchordia**
## 1. Overview

Adding a new delegation strategy with test cases based on mainnet.

## 2. Implementation details

- It is an overall new way for distributing tokens amongst all validators during staking and rewards epoch.
- Not ye integrated with old code.
- Contains an exact main net scenario in test file.

## 3. How to test/use

Run the added test case in the delegation_strategy_test.go file.

## 4. Checklist

- [ ] Does the Readme need to be updated?

### CODE REVIEW

File: x/lscosmos/keeper/delegation_strategy.go [hunk 1]
```diff
@@ -300,3 +300,123 @@ func (k Keeper) GetAllValidatorsState(ctx sdk.Context, denom string) (types.Allo
...
+		},
+	)
+
+	// this is the first round of distribution to all the validators who are not
+	// yet matched to the ideal distribution
+	finalDistribution := make(map[string]sdk.Coin)
+	for _, i := range differenceCurrentAndIdealAmounts {
+
+		if i.Diff.GT(sdk.ZeroDec()) {
+			if newAmount.Amount.ToDec().LT(i.Diff.Abs()) && newAmount.Amount.Sub(newAmount.Amount).GTE(sdk.OneInt()) {
```
Comment (puneet2019) [inline x/lscosmos/keeper/delegation_strategy.go:336]: `if i.Diff.GT(sdk.ZeroDec())` already checks if diff is positive, why need 
`i.Diff.Abs()`

Comment (puneet2019) [reply x/lscosmos/keeper/delegation_strategy.go:336]: `newAmount.Amount.Sub(newAmount.Amount)`
shouldn't this always be zero?

File: x/lscosmos/keeper/delegation_strategy.go [hunk 2]
```diff
@@ -300,3 +300,123 @@ func (k Keeper) GetAllValidatorsState(ctx sdk.Context, denom string) (types.Allo
...
+			Amount:        value,
+		})
+	}
+
+	// sort the types.ValAddressAmounts based on the validator addresses
+	sort.SliceStable(finalValAddressAmounts, func(i, j int) bool {
+		return finalValAddressAmounts[i].ValidatorAddr < finalValAddressAmounts[j].ValidatorAddr
+	})
+
+	// this is a boundary case when there is still some residue left after
```
Comment (puneet2019) [inline x/lscosmos/keeper/delegation_strategy.go:374]: in this case, if you are starting with 0, and there is 10K delegations, will it mean all the 10K go to single validator?

File: x/lscosmos/keeper/delegation_strategy.go (hunk 1 above)
Comment (arhamchordia) [reply x/lscosmos/keeper/delegation_strategy.go:336]: > `if i.Diff.GT(sdk.ZeroDec())` already checks if diff is positive, why need
> `i.Diff.Abs()`

Yes. Makes sense. Will remove this condition.
I think I did it in a haste.

Comment (arhamchordia) [reply x/lscosmos/keeper/delegation_strategy.go:336]: > `newAmount.Amount.Sub(newAmount.Amount)`
> shouldn't this always be zero?

Yeah. Need to subtract it for the next loop for further distribution.

**Pull request closed without merging**